[SLP][NFC]Extract values state/operands analysis into separate class

alexey-bataev opened the pull request:
Extract values state and operands analysis/building into a separate
class. This class allows to localize instrutions state and operands
building for future support of copyable elements vectorization.

alexey-bataev committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.5

alexey-bataev commented:
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> You can test this locally with the following command:
> View the diff from clang-format here.

Clang-format is not related to the change itself, but to the original formatting, need to reformat the whole switch statement

HanKuanChen commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2896,11 +2885,17 @@ class BoUpSLP {
         // operations or alternating sequences (e.g., +, -), we can safely tell
         // the inverse operations by checking commutativity.
-        auto [SelectedOp, Ops] = convertTo(cast<Instruction>(VL[Lane]), S);
+        auto *I = dyn_cast<Instruction>(VL[Lane]);
+        if (!I && isa<PoisonValue>(VL[Lane])) {
```
Can VL[Lane] have a value other than Instruction or PoisonValue?

alexey-bataev replied:
Currently no, in future patches - yes

HanKuanChen replied:
Can we keep
```
        assert((isa<Instruction>(V) || isa<PoisonValue>(V)) &&
               "Expected instruction or poison value");
```
and use `if (isa<PoisonValue>(VL[Lane])) {` here? Also `auto [SelectedOp, Ops] = convertTo(cast<Instruction>(VL[Lane]), S);`.

alexey-bataev replied:
Why? The logic I introduced here does not brake any assumptions. Added the assertion, that the instruction is expected.

HanKuanChen replied:
I just feel it is weird because the following code is removed; however, it could have been kept.
```
        assert((isa<Instruction>(V) || isa<PoisonValue>(V)) &&
               "Expected instruction or poison value");
```
If we don't introduce any new concept, why do we remove the assert?
And if we keep the original assert, we can use
```
        if (isa<PoisonValue>(VL[Lane])) {
          for (unsigned OpIdx : seq<unsigned>(NumOperands))
            OpsVec[OpIdx][Lane] = {Operands[OpIdx][Lane], true, false};
          continue;
        }
        auto [SelectedOp, Ops] = convertTo(cast<Instruction>(VL[Lane]), S);
```

HanKuanChen commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -9734,16 +9725,201 @@ bool BoUpSLP::canBuildSplitNode(ArrayRef<Value *> VL,
+      Type *VL0Ty = VL0->getOperand(IndexIdx)->getType();
+      Type *Ty = all_of(VL,
+                        [VL0Ty](Value *V) {
+                          auto *GEP = dyn_cast<GetElementPtrInst>(V);
+                          if (!GEP)
+                            return true;
+                          return VL0Ty == GEP->getOperand(IndexIdx)->getType();
```
```
if (auto *GEP = dyn_cast<GetElementPtrInst>(V))
  return VL0Ty == GEP->getOperand(IndexIdx)->getType();
return true;
```

alexey-bataev replied:
Reworked it in a bit different way to avoid having 2 returns. This code is a copy-paste from the original code, I did not change anything here

HanKuanChen commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -9734,16 +9725,201 @@ bool BoUpSLP::canBuildSplitNode(ArrayRef<Value *> VL,
+                                    const TargetLibraryInfo &TLI)
+      : DT(DT), DL(DL), TTI(TTI), TLI(TLI) {}
+
+  InstructionsState buildInstructionsState(ArrayRef<Value *> VL) {
+    InstructionsState S = getSameOpcode(VL, TLI);
+    return S;
+  }
```
Do we really need this function?

1. Since the callers already have TLI, they could simply call getSameOpcode themselves, which might be more straightforward.
2. This function only depends on TLI, which is already available to the callers. I don't think we need a wrapper here.

alexey-bataev replied:
Currently, it is just a wrapper around getSameOpcode. But it will be extended later to support "copyable" elements

HanKuanChen replied:
Can we remove it first and add it back with the next patch (copyable elements)?

alexey-bataev replied:
Done

alexey-bataev committed:
address comments

Created using spr 1.3.5

alexey-bataev committed:
address comments

Created using spr 1.3.5

alexey-bataev committed:
address comments

Created using spr 1.3.5

alexey-bataev committed:
Rebase

Created using spr 1.3.5

asb commented:
I've just reverted this commit in 6a2a8ebe27c1941f5b952313239fc6d155f58e9d.

It's causing the RVV CI to fail with a crash compiling oggenc.c from llvm-test-suite. I've reduced the C input to come up with this test case:

reduced oggenc.c:
```c
float *mdct_forward_x;
int mdct_forward_j;
void mdct_butterfly_16(float *x) {
  float r0 = x[6] + 2, r1 = x[6] - x[2], r2 = 4 + x[0], r3 = x[4] - x[0];
  x[4] = r0 - r2;
  r0 = 5 - x[1];
  r2 = 7 - x[3];
  x[0] = r0;
  x[2] = r1 - r0;
  x[3] = r2 + r3;
  x[5] = r1 - r0;
}
void mdct_butterfly_32(float *x) { mdct_butterfly_16(x + 6); }
void mdct_forward() {
  for (;;)
    mdct_butterfly_32(mdct_forward_x + mdct_forward_j);
}
```

Then to reproduce the crash:
```
~/llvm-project/build/release/bin/clang --target=riscv64-linux-gnu -march=rv64gcv -O3 -c oggenc.c -fno-strict-aliasing
clang: ../../llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:19512: auto llvm::slpvectorizer::BoUpSLP::BlockScheduling::tryScheduleBundle(ArrayRef<Value *>, BoUpSLP *, const InstructionsState &)::(anonymous class)::operator()(bool, ScheduleBundle &) const: Assertion `Picked->isReady() && "must be ready to schedule"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/asb/llvm-project/build/release/bin/clang --target=riscv64-linux-gnu -march=rv64gcv -O3 -c oggenc.c -fno-strict-aliasing
1.	<eof> parser at end of file
2.	Optimizer
3.	Running pass "function<eager-inv>(float2int,lower-constant-intrinsics,chr,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O3>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "oggenc.c"
4.	Running pass "slp-vectorizer" on function "mdct_forward"
 #0 0x000060a794223306 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/asb/llvm-project/build/release/bin/clang+0x8789306)
 #1 0x000060a794220c1e llvm::sys::RunSignalHandlers() (/home/asb/llvm-project/build/release/bin/clang+0x8786c1e)
 #2 0x000060a79418239d CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007c6e1424c1d0 (/usr/lib/libc.so.6+0x3d1d0)
 #4 0x00007c6e142a5334 (/usr/lib/libc.so.6+0x96334)
 #5 0x00007c6e1424c120 raise (/usr/lib/libc.so.6+0x3d120)
 #6 0x00007c6e142334c3 abort (/usr/lib/libc.so.6+0x244c3)
 #7 0x00007c6e142333df (/usr/lib/libc.so.6+0x243df)
 #8 0x00007c6e14244177 (/usr/lib/libc.so.6+0x35177)
 #9 0x000060a795db0d31 llvm::slpvectorizer::BoUpSLP::BlockScheduling::tryScheduleBundle(llvm::ArrayRef<llvm::Value*>, llvm::slpvectorizer::BoUpSLP*, (anonymous namespace)::InstructionsState const&)::$_0::operator()(bool, llvm::slpvectorizer::BoUpSLP::ScheduleBundle&) const SLPVectorizer.cpp:0:0
#10 0x000060a795d1b82a llvm::slpvectorizer::BoUpSLP::buildTreeRec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/home/asb/llvm-project/build/release/bin/clang+0xa28182a)
#11 0x000060a795d1ad91 llvm::slpvectorizer::BoUpSLP::buildTreeRec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/home/asb/llvm-project/build/release/bin/clang+0xa280d91)
........
........
```

Looking at the diff of a log with `-mllvm -debug -mllvm -print-after-all` I see it starts to diverge in the ordering of some operand ordering in printouts (i.e. under `Operand 0:`).

alexey-bataev commented:
> I've just reverted this commit in [6a2a8eb](https://github.com/llvm/llvm-project/commit/6a2a8ebe27c1941f5b952313239fc6d155f58e9d) (EDIT: and here's the build going green again after this revert https://lab.llvm.org/buildbot/#/builders/132/builds/1052)
> 
> It's causing the RVV CI to fail with a crash compiling oggenc.c from llvm-test-suite. I've reduced the C input to come up with this test case:
> 
> reduced oggenc.c:
> 
> ```c
> float *mdct_forward_x;
> int mdct_forward_j;
> void mdct_butterfly_16(float *x) {
>   float r0 = x[6] + 2, r1 = x[6] - x[2], r2 = 4 + x[0], r3 = x[4] - x[0];
>   x[4] = r0 - r2;
>   r0 = 5 - x[1];
>   r2 = 7 - x[3];
>   x[0] = r0;
>   x[2] = r1 - r0;
>   x[3] = r2 + r3;
>   x[5] = r1 - r0;
> }
> void mdct_butterfly_32(float *x) { mdct_butterfly_16(x + 6); }
> void mdct_forward() {
>   for (;;)
>     mdct_butterfly_32(mdct_forward_x + mdct_forward_j);
> }
> ```
> 
> Then to reproduce the crash:
> 
> ```
> ~/llvm-project/build/release/bin/clang --target=riscv64-linux-gnu -march=rv64gcv -O3 -c oggenc.c -fno-strict-aliasing
> clang: ../../llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:19512: auto llvm::slpvectorizer::BoUpSLP::BlockScheduling::tryScheduleBundle(ArrayRef<Value *>, BoUpSLP *, const InstructionsState &)::(anonymous class)::operator()(bool, ScheduleBundle &) const: Assertion `Picked->isReady() && "must be ready to schedule"' failed.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
> Stack dump:
> 0.	Program arguments: /home/asb/llvm-project/build/release/bin/clang --target=riscv64-linux-gnu -march=rv64gcv -O3 -c oggenc.c -fno-strict-aliasing
> 1.	<eof> parser at end of file
> 2.	Optimizer
> 3.	Running pass "function<eager-inv>(float2int,lower-constant-intrinsics,chr,loop(loop-rotate<header-duplication;no-prepare-for-lto>,loop-deletion),loop-distribute,inject-tli-mappings,loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>,infer-alignment,loop-load-elim,instcombine<max-iterations=1;no-verify-fixpoint>,simplifycfg<bonus-inst-threshold=1;forward-switch-cond;switch-range-to-icmp;switch-to-lookup;no-keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,slp-vectorizer,vector-combine,instcombine<max-iterations=1;no-verify-fixpoint>,loop-unroll<O3>,transform-warning,sroa<preserve-cfg>,infer-alignment,instcombine<max-iterations=1;no-verify-fixpoint>,loop-mssa(licm<allowspeculation>),alignment-from-assumptions,loop-sink,instsimplify,div-rem-pairs,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;speculate-unpredictables>)" on module "oggenc.c"
> 4.	Running pass "slp-vectorizer" on function "mdct_forward"
>  #0 0x000060a794223306 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/asb/llvm-project/build/release/bin/clang+0x8789306)
>  #1 0x000060a794220c1e llvm::sys::RunSignalHandlers() (/home/asb/llvm-project/build/release/bin/clang+0x8786c1e)
>  #2 0x000060a79418239d CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
>  #3 0x00007c6e1424c1d0 (/usr/lib/libc.so.6+0x3d1d0)
>  #4 0x00007c6e142a5334 (/usr/lib/libc.so.6+0x96334)
>  #5 0x00007c6e1424c120 raise (/usr/lib/libc.so.6+0x3d120)
>  #6 0x00007c6e142334c3 abort (/usr/lib/libc.so.6+0x244c3)
>  #7 0x00007c6e142333df (/usr/lib/libc.so.6+0x243df)
>  #8 0x00007c6e14244177 (/usr/lib/libc.so.6+0x35177)
>  #9 0x000060a795db0d31 llvm::slpvectorizer::BoUpSLP::BlockScheduling::tryScheduleBundle(llvm::ArrayRef<llvm::Value*>, llvm::slpvectorizer::BoUpSLP*, (anonymous namespace)::InstructionsState const&)::$_0::operator()(bool, llvm::slpvectorizer::BoUpSLP::ScheduleBundle&) const SLPVectorizer.cpp:0:0
> #10 0x000060a795d1b82a llvm::slpvectorizer::BoUpSLP::buildTreeRec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/home/asb/llvm-project/build/release/bin/clang+0xa28182a)
> #11 0x000060a795d1ad91 llvm::slpvectorizer::BoUpSLP::buildTreeRec(llvm::ArrayRef<llvm::Value*>, unsigned int, llvm::slpvectorizer::BoUpSLP::EdgeInfo const&, unsigned int) (/home/asb/llvm-project/build/release/bin/clang+0xa280d91)
> ........
> ........
> ```
> 
> Looking at the diff of a log with `-mllvm -debug -mllvm -print-after-all` I see it starts to diverge in the ordering of some operands listed in printouts (i.e. under `Operand 0:`).

Thanks, fixed
